openvdb/points/AttributeArray.h

```diff
@@ -169,6 +205,37 @@ class OPENVDB_API AttributeArray
     /// @note This counts each data element in a strided array
     virtual Index dataSize() const = 0;
 
+#if OPENVDB_ABI_VERSION_NUMBER >= 6
+    /// Return the name of the value type of a single element in this array (e.g., "float" or "vec3d").
+    virtual Name valueType() const = 0;
+
+    /// Return the name of the codec used by this array (e.g., "trnc" or "fxpt").
+    virtual Name codecType() const = 0;
+
+    /// Return the size in bytes of the value type of a single element in this array.
+    /// (e.g. "float" -> 4 bytes, "vec3d" -> 24 bytes").
+    virtual Index valueTypeSize() const = 0;
+
+    /// Return the size in bytes of the storage type of a single element of this array.
+    /// @note If the Codec is a NullCodec, valueSize() == storageSize()
+    virtual Index storageTypeSize() const = 0;
+
+    /// Return @c true if the value type is floating point
+    virtual bool valueTypeIsFloatingPoint() const = 0;
```

Comment:
No, this PR doesn't rely on that change. I made these two changes orthogonal, so they wouldn't have to be merged together. If there's no consensus for accepting the other PR, I can always make a new change to introduce those traits into a local namespace of this class to speed up this slow runtime string comparison.

Reply:
I'm referring explicitly to the functionality of this particular function - ```VecTraits<ValueType>::ElementType``` will work correctly for scalars and vectors but quats and matrices will behave unexpectedl i.e. float = true, vec3f = true, mat3f = false

Reply:
Actually, that's not true and this specific case is tested in the unit tests - https://github.com/AcademySoftwareFoundation/openvdb/pull/281/files#diff-09ba2ff10db2a9aef3551d5ca873020dR537.

What you're saying is almost correct though in that a mat3i type would fail because the std::is_integral test here would resolve to using the struct not the element type. I'm ignoring this particular case because we don't use integer matrices or quaternions. However were the traits PR to go in, I'd update this method to use that instead so it's more correct in handling this edge case.

Reply:
Added a comment to this effect in the code.

Reply:
Ah I assumed you were using is_floating_point not is_integral. I think you should be using is_floating_point due to the nature of this function and the expected result. With all these methods, it's better practise to implement what they are actually defining instead of what they aren't e.g. what If i want a pointer array (float*) or string array (I realise strings are handled differently but there's not reason why you shouldn't be able to create one). These would return true!

If you're worried about supporting half, just add as is_same\<half>. For the rest, I'd advise adding an is_same<> for mat and quat fp types until your value traits PR is in (with a comment).

Reply:
I've tweaked the logic here to match what you're suggesting. Bear in mind that the most important aspect of this PR is adding the virtual function, not the actual implementation itself which we can continue to improve later. I considered simply throwing NotImplementedErrors as the 6.0.0 change and returning to do the implementation later, but felt it was simple enough to at least get a first version implementation in place.

Reply:
Yeah sure I get that, that would have been fine also imo. This looks good now!